contracts/utility/MathEx.sol

```diff
@@ -158,4 +159,104 @@ library MathEx {
+    */
+    function mulDivF(uint256 x, uint256 y, uint256 z) internal pure returns (uint256) {
+        (uint256 xyh, uint256 xyl) = mul512(x, y);
+        if (xyh == 0) { // `x * y < 2 ^ 256`
+            return xyl / z;
+        }
+        if (xyh < z) { // `x * y / z < 2 ^ 256`
+            uint256 m = mulMod(x, y, z);                    // `m = x * y % z`
+            (uint256 nh, uint256 nl) = sub512(xyh, xyl, m); // `n = x * y - m` hence `n / z = floor(x * y / z)`
+            if (nh == 0) { // `n < 2 ^ 256`
+                return nl / z;
+            }
+            uint256 p = unsafeSub(0, z) & z; // `p` is the largest power of 2 which `z` is divisible by
+            uint256 q = div512(nh, nl, p);   // `n` is divisible by `p` because `n` is divisible by `z` and `z` is divisible by `p`
+            uint256 r = inv256(z / p);       // `z / p = 1 mod 2` hence `inverse(z / p) = 1 mod 2 ^ 256`
+            return unsafeMul(q, r);          // `q * r = (n / p) * inverse(z / p) = n / z`
+        }
+        revert("ERR_OVERFLOW"); // `x * y / z >= 2 ^ 256`
+    }
+
+    /**
+      * @dev Compute the smallest integer larger than or equal to `x * y / z`
+    */
+    function mulDivC(uint256 x, uint256 y, uint256 z) internal pure returns (uint256) {
+        uint256 w = mulDivF(x, y, z);
+        if (mulMod(x, y, z) > 0) {
+            require(w < MAX_UINT256, "ERR_OVERFLOW");
+            return w + 1;
+        }
+        return w;
+    }
+
+    /**
+      * @dev Compute the value of `x * y`
+    */
+    function mul512(uint256 x, uint256 y) private pure returns (uint256, uint256) {
+        uint256 p = mulModMax(x, y);
+        uint256 q = unsafeMul(x, y);
+        if (p >= q) {
+            return (p - q, q);
+        }
+        return (unsafeSub(p, q) - 1, q);
+    }
+
+    /**
+      * @dev Compute the value of `2 ^ 256 * xh + xl - y`, where `2 ^ 256 * xh + xl >= y`
+    */
+    function sub512(uint256 xh, uint256 xl, uint256 y) private pure returns (uint256, uint256) {
+        if (xl >= y) {
+            return (xh, xl - y);
+        }
+        return (xh - 1, unsafeSub(xl, y));
+    }
+
+    /**
+      * @dev Compute the value of `(2 ^ 256 * xh + xl) / pow2n`, where `xl` is divisible by `pow2n`
+    */
+    function div512(uint256 xh, uint256 xl, uint256 pow2n) private pure returns (uint256) {
+        uint256 pow2nInv = unsafeAdd(unsafeSub(0, pow2n) / pow2n, 1);
+        return unsafeMul(xh, pow2nInv) | (xl / pow2n);
+    }
+
+    /**
+      * @dev Compute the inverse of `x` modulo `2 ^ 256`, where `x` is congruent to `1` modulo `2`
+    */
+    function inv256(uint256 x) private pure returns (uint256) {
+        uint256 inv = 1;
+        for (uint256 i = 0; i < 8; i++) {
+            inv = unsafeMul(inv, unsafeSub(2, unsafeMul(inv, x)));
+        }
+        return inv;
+    }
+
+    // does not revert on overflow
+    function unsafeAdd(uint256 x, uint256 y) private pure returns (uint256) {
```

Comment:
Why do we need these `unsafeXXX` methods explicitly? Are you preparing to use `unchecked` in 0.8? 😉

Reply:
Yes, but not only.

Mainly for two reasons:
1. As you've mentioned - when porting to solc 0.8, we will need to scope every piece of code which allows arithmetic overflow with `unchecked`.
2. It will make it easier for us to easily emulate this entire module in "natural" languages (for testing purposes), where overflows do not occur. In other words, we will have all the "overflowing" stuff concentrated in a single set of functions rather than being scattered all over.

For example, in Python you'd do:
```
MAX_VAL = (1 << 256) - 1

def unsafeAdd(x, y): return (x + y) & MAX_VAL
def unsafeSub(x, y): return (x - y) & MAX_VAL
def unsafeMul(x, y): return (x * y) & MAX_VAL
```

Reply:
👍